processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -25,26 +25,29 @@
 import io.druid.data.input.Row;
 import io.druid.query.filter.DimFilter;
 import io.druid.query.filter.ValueMatcher;
-import io.druid.query.groupby.RowBasedValueMatcherFactory;
+import io.druid.query.groupby.RowBasedColumnSelectorFactory;
+import io.druid.segment.column.ValueType;
 
 import java.nio.ByteBuffer;
+import java.util.Map;
 
-public class DimFilterHavingSpec implements HavingSpec
+public class DimFilterHavingSpec extends BaseHavingSpec
 {
   private static final byte CACHE_KEY = (byte) 0x9;
 
   private final DimFilter dimFilter;
-  private final RowBasedValueMatcherFactory valueMatcherFactory;
-  private final ValueMatcher valueMatcher;
+  private final ThreadLocal<Row> rowSupplier;
+
+  private final Object valueMatcherMonitor = new Object();
+  private volatile ValueMatcher valueMatcher;
```

Comment:
@gianm not only that, I pointed out that making `valueMatcher` `volatile` doesn't really provide you any extra guarantees. To guarantee visibility of some value in some field, you should add `synchronized`, `volatile`, or some other HB primitives *on something else*, not on the field itself.